locoprop/layer.py

```diff
@@ -1,19 +1,106 @@
+        return x
+
+    @staticmethod
+    def backward(ctx, dy: torch.Tensor):
+        return dy, dy
+
+
+def maybe_detach(x: Any):
+    if isinstance(x, torch.Tensor):
+        return x.detach().requires_grad_(True)
+    return x
+
+
 class LocoLayer(nn.Module):
     def __init__(self, module, activation, implicit=False):
         super().__init__()
         self.module = module
         self.activation = activation
-        self.implicit = implicit
+        self.storage_for_grad = None
+        self.lctx = LocopropCtx(implicit=implicit)
 
-    def forward(self, x=None, hidden=None):
-        if x is None and hidden is None:
-            raise ValueError("No argument was given. Provide either input or hidden state.")
+    def _training_first_pass(self, *args, **kwargs):
+        hidden = self.module(*args, **kwargs)
+        self.storage_for_grad = torch.empty_like(hidden, requires_grad=True)
+        return CloneGrad.apply(hidden, self.storage_for_grad)
 
-        if hidden is None:
-            hidden = self.module(x)
+    def _training_second_pass(self, *args, **kwargs):
+        dy = self.storage_for_grad.grad
+        self.module.requires_grad_(True)
+        original_params = {n: p.clone() for n, p in self.module.named_parameters()}
 
-        if self.implicit:
+        opt = self.lctx.optimizer
+        base_lrs = [group["lr"] for group in opt.param_groups]
+        opt.zero_grad()
+
+        for i in range(self.lctx.iterations):
+            for base_lr, group in zip(base_lrs, opt.param_groups):
+                group["lr"] = base_lr * max(1.0 - i / self.lctx.iterations, 0.25)
+            with torch.enable_grad():
+                # pre-activation:
+                args = [maybe_detach(a) for a in args]
+                kwargs = {k: maybe_detach(v) for k, v in kwargs.items()}
+                a = self.module(*args, **kwargs)
+                y = self.activation(a)  # <- post-activation
+            if i == 0:
+                with torch.no_grad():
+                    hidden = a.detach()
+                    post_target = (y - self.lctx.learning_rate * dy).detach()
+
+            torch.autograd.backward([y], [(y - post_target) / a.size(0)], inputs=list(self.module.parameters()))
+            opt.step()
+            opt.zero_grad()
+
+        for base_lr, group in zip(base_lrs, opt.param_groups):
+            group["lr"] = base_lr
+
+        with torch.no_grad():
+            for n, p in self.module.named_parameters():
+                p.grad = original_params[n].data - p.data
+                p.set_(original_params[n].data)
+
+        self.storage_for_grad = None
+        if self.lctx.correction <= 0:
             return hidden
+
+        # correct input of next layer
+        with torch.no_grad():
+            delta = self.module(*args, **kwargs) - hidden
+            correction = self.lctx.correction / delta.std().clamp(min=self.lctx.correction_eps)
+            return hidden + correction.clamp(max=1) * delta
```

Comment:
For reference, it used to be:
```python
with torch.no_grad():
    delta = layer(inp) - inps[i + 1]
    norm = delta.norm() + 1e-5
    inps[i + 1] += min(norm, self.correction * delta.size(1) ** 0.5) * delta / norm
```

Reply:
`min(norm, x) / norm = min(1, x / norm) = (x / norm).clamp(max=1)`
with `x = self.correction * delta.size(1) ** 0.5` and `norm = delta.norm()`
`(x / norm).clamp(max=1) = (self.correction / delta.std()).clamp(max=1)`
for stability purposes, we can add an epsilon to the std. Here, I simply prefer clamping epsilon in place rather than adding it, so that 1e-5 won't be turned into 2e-5.

Reply:
Not sure what you mean in the first line: Suppose `norm = 2` and `x = 1`, then `min(norm, x) = 1` and `min(1, x / norm) = 0.5`.

For the second line, you're essentially saying that `delta.norm() / sqrt(dim) = delta.std()` with `dim = delta.size(1)` which I don't see where it's coming from. Suppose `delta = torch.ones(...)`, then `delta.std() = 0` and `delta.norm() / sqrt(dim) = 1`. Am I missing something?

Reply:
`min(norm, x) = 1`
but you'd still divide by norm afterwards, resulting in 0.5

you're right, the second line is wrong. it'd only be correct if it was delta.numel()

Reply:
Oh right, missed the division by norm. You're absolutely right.

Second point: Even still, the result is the same for e.g. `delta = torch.ones(8)`. Here, `delta.numel() = delta.size(0)`. There might be some merit to the std-based approach, but I don't see immediately why it should work (better).

Reply:
should it also be delta.size(1) for convnets, or would `delta.numel() / delta.size(0)` be better?

Reply:
> why it should work (better).
purely code simplification. norm is likely more stable, as the epsilon is in front of the sqrt.

Reply:
Good point, depends on the behavior of `norm()`. Thinking about it, it probably always should have been `norm = delta.view(delta.size(0), -1).norm(1).mean()` and `(delta.numel() / delta.size(0))**0.5`. The idea is to account for growing norm magnitude with growing dimensionality. Not sure how sound this is to begin with but it seemed more stable than just the vanilla norm when I tried out both.

Reply:
Or probably even `delta.view(delta.size(0), -1).norm(1)` to have per-sample corrections, which makes much more sense.

Reply:
wouldn't `x.flatten(1).norm(1) * (x.numel() / x.size(0))**0.5` be equal to `x.flatten(1).std(1)`?

Reply:
So IIRC we have `std(x) = ((x - x.mean()).pow(2).sum() / (x.numel() - 1)).sqrt()` whereas `x.norm() = x.pow(2).sum().sqrt()`. Standard deviation quantifies the variance in the data and depends on `x`, the dimension normalizer is independent of data. (consider the case where `x = torch.ones(8)`)

Reply:
Tested the two approaches like this:
```python
x0, x1 = torch.ones(16, 32, 32, 3), torch.zeros(16, 32, 32, 3)
correction = 0.1

delta = x0 - x1
mag = correction / delta.flatten(1).std(1).clamp(min=1e-5).clamp(max=1)
delta = (mag.unsqueeze(-1) * delta.flatten(1)).view(*delta.shape)
res1 = x1 + delta

delta = x0 - x1
norm = delta.flatten(1).norm(dim=1).clamp(min=1e-5)
mag = (correction / norm * delta.size(1) ** 0.5).clamp(max=1)
delta = (mag.unsqueeze(-1) * delta.flatten(1)).view(*delta.shape)
res2 = x1 + delta

(res1 - res2).norm().sum()
>>> tensor(2217092.7500)
```

Going to reintroduce the second approach because I don't understand the logic behind the first one (it might still be justified, I just can't see it right now). We should be good to merge then.

Reply:
You're right. My assumption was that std = norm*constant, which is only correct of it's already centered.

Reply:
Ah I see what you meant now!

Reply:
I fused the .min() and .max() calls into one. Probably good if you have another look before merging.